spec/dynamic/layout.md

```diff
@@ -0,0 +1,333 @@
+
+[§](r[dynamic.layout.pointer.wide]) A pointer to a `!Thin` type has both a data portion and a metadata portion. 
+
+[§](r[dynamic.layout.pointer.wide-underlying]) The wide pointer type `*mut T` has an underlying type of the given *exposition only* definition, where `M` is the pointer metadata type of `T`. The wide pointer type `*const T` has the same underlying type
+```rust
+struct WidePtr{
+    data: *mut (),
+    metadata: M
+}
+```
+
+[§](r[dynamic.layout.pointer.wide-validity]) A given value of a pointer type is valid if the data portion is valid and the metadata portion is valid.
+
+[§](r[dynamic.layout.pointer.ref]) The reference type `&T` has an underlying type of `*const T`. The reference type `&mut T` has an underlying type of `*mut T`.
+
+[§](r[dynamic.layout.pointer.ref-validity]) A value of a reference type is valid if the pointer value has an address which is not the value `0` and is a multiple of the dynamic alignment requirement of the value.
+
+[!NOTE]: The dynamic alignment requirement is the alignment requirement of the type if the type has one, or otherwise if the pointer metadata type is a trait object vtable pointer, the alignment requirement accessible from the pointer metadata. A pointer with the address `0` is known as a null pointer.
+
+[§](r[dynamic.layout.pointer.fn-size-align]) A function pointer type has an target dependant size and alignment requirement. The size and alignment does not depend on the parameters, return type, or ABI tag of the function pointer type.
+
+[!NOTE]: This is typically, but not always, the same as the size and alignment requirement of `*const ()`.
+
+[§](r[dynamic.layout.pointer.fn-repr]) The address of a function pointer value is the same as the value of type `usize` computed from the same bytes. The pointer tag of a pointer value is each pointer portion of each byte of the representation. 
+
+[§](r[dynamic.layout.pointer.fn-validity]) A value of an function pointer type is valid if the address of the pointer is not `0`.
+
+## Aggregate Layout [§](r[dynamic.layout.aggregate])
+
+[§](r[dynamic.layout.aggregate.fields]) Each aggregate type has a list of fields, that each have a type and offset. 
+
+[!NOTE]: This includes built-in tuple types and arrays. This does not include `enum` definitions.
+
+[§](r[dynamic.layout.aggregate.struct-base] The fields of a `struct` definition are each layed at offsets that ensure that each field occupies nonoverlapping storage, which, unless modified by the `repr(packed)` attribute (r#[dynamic.layout.aggregate.packed])) is suitably aligned. The size of a `struct` definition is at least sufficient to store each field, and the alignment requirement of a `struct` definition, unless modified by the `repr(packed)` attribute, is suitable to align each field.
+
+[!NOTE]: The offsets of the fields, sizes, and alignment requirements of two different `struct` types, even with the same field types in the same order, may be different.
```

Comment:
maybe say "same source order" here? Not sure if that's an existing concept

> even with the same field types in the same declaration order, may be different unless modified by the `repr(C)` attribute.

Reply:
Source order is probably fine since it's it's a note.